### ISSUE TITLE
use yaml for proto.lock to reduce diff

### ISSUE THREAD

**Issue by craynic** [nilslice/protolock#115]
We are managing hundreds of protos which result in a lock file with one hundred thousand lines. When we changed something, and commit the changes to the lock, a very large diff will be produced with many many lines changed, though in fact I only added a few files or fields. I think this is because the json file has many braces. The braces always match in diff, causing diff believes that the paths or the field names have changed. I think yaml shoul be a better choice for this lock file, which has no braces in independent lines, and it can store small objects in one line (so that there won't be many `id: 0`).

**Comment by nilslice** [comment]
@craynic - thank you for the suggestion! I too manage a project (for which this tool was created initially) that has a similar scale of protos to track. While YAML could be an alternative, I don't think the resulting "win" would be significant enough to make this kind of change. Truthfully, you should not be concerned with the lock file 99% of the time - just let `protolock` manage it and trust that it will do the right thing. 

I had considered using protobuf (in #34) as an alternate format, which would be more compact, but decided against it. If you wanted to try a proof of concept using YAML and submit a PR, I'd be happy to check it out, but would need to support plugins and pass all CI tests before thinking about supporting it.

**Comment by craynic** [comment]
Well, yaml may not be the best choice, but should be better than json.
For now, making the lock file in yaml may bring some benefits, but it will also bring break changes. I truely doubt if the benefits worth the changes. Maybe a compatibility layer is needed to let the old json lock file work.
For proto-buf, I don't think it's a good option. Most CVS track files by line, but proto files behaves like a binary file. A little change may cause the whole file to be different.
I'm willing to do the job, but I'm not familiar with golang, so that may take some time.

**Comment by nilslice** [comment]
@craynic - I appreciate the thought you've put into this, but after some additional review, I've decided to not consider a lock file format change. I only want to keep you informed so you don't do the work and anticipate the feature merged.

As your project continues to grow, please let me know if there are other issues. Also, ensure you are using the latest versions. There was a feature added a while ago to add deterministic ordering of the lock file, so that diffs were easier to read -- maybe this will help you if you don't already have this version installed. 

Thanks!

**Comment by craynic** [comment]
Could you please provide the commit id for the feature about deterministic ordering? We've worked with many versions of protolock.

**Comment by nilslice** [comment]
@craynic - it looks like this was the release: 
https://github.com/nilslice/protolock/releases/tag/v0.10.0

**Comment by fenollp** [comment]
I too am looking for a way to have turn today's JSON lockfile into YAML mainly for quicker diff reviews.
There're talk about potential breaking changes in this thread and I fail to see any given:
* these lockfiles are machine generated so should always have a pretty regular layout / not use "weird" features of either formats
* JSON is a subset of YAML
* https://github.com/ghodss/yaml is a drop-in replacement for de/serialization of JSON/YAML so both formats can be supported at the same time

ghodss' YAML support for ordering fields may be dodgy still as it still relies on go-yaml v2 when v3 introduced it (to confirm).

In conclusion: I'd like to reopen this issue and use ghodss' lib and introduce a flag that would switch the output format of lockfiles.
